### PR TITLE
Enable client-side rate limiting on source-stripe #31512

### LINKED ISSUE #31512: Enable client-side rate limiting on source-stripe
## What

As a follow up to https://github.com/airbytehq/airbyte/issues/29628, let's enable rate limiting on the Stripe source.

The rate limit should depend on the type of account used.
- test accounts have a rate limit of 25 requests per second
  - test accounts have an client_secret that starts with "sk_test_"
- prod accounts have a rate limit of 100 rps (to be confirmed)
- It should also be possible for the user to configure a rate limit
  - Through the config with a default parameter (None or 0 means use the account's limit)
    -  can be an hidden flag

## Acceptance criteria
- [ ] The stripe connector respects a configured rate limit
- [ ] Test and prod accounts have separate rate limits
- [ ] A user can configure the rate limit applied to its connection
- [ ] The rate limit is applied both to "Stream" classes and concurrent classes

### PR DESCRIPTION
**Pull request by keu**
## What

This adds client-side rate limiting that helps increase the number of workers and not hit the call rate limit when it is too big.

## How
*Describe the solution*

## Recommended reading order
1. `spec.yaml`
2. `source.py`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/resources/pull-requests-handbook)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:generateScaffolds` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

<details><summary><strong>Updating the Python CDK</strong></summary>

### Airbyter

Before merging:
- Pull Request description explains what problem it is solving
- Code change is unit tested
- Build and my-py check pass
- Smoke test the change on at least one affected connector
   - On Github: Run [this workflow](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml), passing `--use-local-cdk --name=source-<connector>` as options
   - Locally: `airbyte-ci connectors --use-local-cdk --name=source-<connector> test`
- PR is reviewed and approved
      
After merging:
- [Publish the CDK](https://github.com/airbytehq/airbyte/actions/workflows/publish-cdk-command-manually.yml)
   - The CDK does not follow proper semantic versioning. Choose minor if this the change has significant user impact or is a breaking change. Choose patch otherwise.
   - Write a thoughtful changelog message so we know what was updated.
- Merge the platform PR that was auto-created for updating the Connector Builder's CDK version
   - This step is optional if the change does not affect the connector builder or declarative connectors.

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-stripe/source_stripe/source.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-stripe/source_stripe/source.py:163]: Should this be `api_budget` instead of `call_budget`? I fear this code would lead to `unexpected keyword argument` See: https://github.com/airbytehq/airbyte/blob/7d5026405a2b6ecfd8f0fa99166f91c8fa93db82/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py#L42

File: airbyte-integrations/connectors/source-stripe/source_stripe/source.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-stripe/source_stripe/source.py:125]: Should we have `call_limit = max_call_rate` in that case?

**Review by maxi297 — CHANGES_REQUESTED** [review]
Can we add some test to ensure that we get the behavior we want?

File: airbyte-integrations/connectors/source-stripe/source_stripe/source.py
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/source_stripe/source.py:125]: Adding more information: If `config["call_rate_limit"] = 200`, it seems like the `FixedWindowCallRatePolicy` without matchers will have a call_limit of 200 which is bigger than the max allowed

Comment (keu) [reply airbyte-integrations/connectors/source-stripe/source_stripe/source.py:125]: right, I absolutely missed this part 
actually it is in in another change set merged with this PR https://github.com/airbytehq/airbyte/pull/31696
I'm waiting for it to be merged to update this.

**Comment by keu** [comment]
blocked by this https://github.com/airbytehq/airbyte/pull/31696/files?diff=unified&w=0
I'm waiting for it to update this PR with logic for concurrent streams.

**Comment by keu** [comment]
A comment about choosing MovingWindow over FixedWindow policy.
The initial assumption was that on such a small limit interval as 1 second, the bucket strategy would underperform and would be close to FixedWindow.
The testing was done with the level of concurrency = 15.
The results:

| policy       | sync time                                    | sleep calls | API limit hits |
|--------------|----------------------------------------------|-------------|----------------|
| MovingWindow | 4.84s user 0.72s system 7% cpu 1:18.81 total | 152         | 0              |
| FixedWindow  | 5.97s user 0.85s system 7% cpu 1:33.92 total | 35          | 4              |

**Review by maxi297 — COMMENTED** [review]
Thanks for adding the tests! We just need to do the items from the connector checklist and we should be good to go. There is this blocked situation with the other PR that I don't fully get though so if you can expand on that just to be sure we're aligned

File: airbyte-integrations/connectors/source-stripe/source_stripe/spec.yaml
Comment (keu) [reply airbyte-integrations/connectors/source-stripe/source_stripe/spec.yaml:79]: I have added this just for your convenience to test with different policies :)

**Review by girarda — COMMENTED** [review]
I ran a test locally with a live account and 15 workers and the moving window strategy.

The acquire_call CPU usage was ~2% which isn't outrageous to me

Screenshot 2023-11-14 at 10 54 43 AM

File: airbyte-integrations/connectors/source-stripe/setup.py
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/setup.py:8]: note: we'll need to bump this to the latest version with the higher retry count

File: airbyte-integrations/connectors/source-stripe/source_stripe/source.py
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/source_stripe/source.py:193]: the default should use a policy

File: airbyte-integrations/connectors/source-stripe/source_stripe/source.py
Comment (keu) [reply airbyte-integrations/connectors/source-stripe/source_stripe/source.py:150]: The changes that @yevhenii-ldv did include moving from HttpStream to HttpRequester which needs to receive call_budget with proper rate limited Session. 
But because his PR is not going anywhere soon, I think this PR is not blocked anymore.

File: airbyte-integrations/connectors/source-stripe/source_stripe/source.py
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/source_stripe/source.py:48]: [the limit in the spec ](https://github.com/airbytehq/airbyte/pull/32284/files#diff-31ac2b459174644b10c8b738b4fe98173c4cdfded04d9db4f65a4cfa67d0ab00R64)should be aligned with this limit

File: airbyte-integrations/connectors/source-stripe/source_stripe/spec.yaml
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/source_stripe/spec.yaml:64]: maximum should be 20

File: airbyte-integrations/connectors/source-stripe/source_stripe/spec.yaml
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/source_stripe/spec.yaml:65]: let's set the default to 10 and go from there

File: airbyte-integrations/connectors/source-stripe/source_stripe/spec.yaml
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/source_stripe/spec.yaml:68]:
```suggestion
        The number of worker thread to use for the sync.
```

Comment (girarda) [reply airbyte-integrations/connectors/source-stripe/source_stripe/spec.yaml:68]: this statement isn't true

File: airbyte-integrations/connectors/source-stripe/source_stripe/source.py
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/source_stripe/source.py:148]: this will never be true since call_rate_limit was removed from the spec

File: airbyte-integrations/connectors/source-stripe/source_stripe/source.py
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/source_stripe/source.py:160]: call_rate_policy isn't in the pec so this will never be true

Comment (keu) [reply airbyte-integrations/connectors/source-stripe/source_stripe/source.py]: I think something went wrong with merge, so this was an old code, updated

File: docs/integrations/sources/stripe.md
Comment (girarda) [inline docs/integrations/sources/stripe.md:219]: nit: let's update the date before merging

**Review by girarda — APPROVED** [review]
:shipit: !

File: docs/integrations/sources/stripe.md
Comment (keu) [reply docs/integrations/sources/stripe.md]: done

**Comment by keu** [comment]
@maxi297 
Screenshot 2023-11-16 at 03 51 26

**Review by maxi297 — APPROVED** [review]
Check list has been updated. I'm good with the implementation

**Pull request merged**
